### PR TITLE
Base implementation of the PyMediaLibrary

### PR DESCRIPTION
**Pull request by geffrak**
Signed-off-by: Kerby Geffrard <kerby.geffrard@autodesk.com>

### Linked issues

N/A

### Summarize your change.

Implements the PyMediaLibrary and capabilities for MediaLibrary plugins in OpenRV.

#### Implementation details

The plugins are lazy loaded up until a media is loaded. They are all stored in the PyRootNode. The PyRootNode then shares the plugin object with the children PyMediaNode. That way, they are only loaded once and the performance stays great.

The plugins are required to implement the MediaAPI. You have an example in the demonstration plugin.

```
{ "is_plugin_enabled" },
{ "is_library_media_url" },
{ "is_streaming" },
{ "is_redirecting" },
{ "get_http_cookies" },
{ "get_http_headers" },
{ "get_http_redirection" }
```

#### Demonstration plugin
Please, don't use it in production. 

It comes with a demonstration plugin that needs to be enabled with the `RV_MEDIA_LIBRARY_UNSECURE_DEMO_ENABLED` set to 1. 

Once enabled, you can set the cookies for one site using `RV_MEDIA_LIBRARY_UNSECURE_DEMO_COOKIES`. It's parsed by Python's SimpleCookie module and it allows you to set multiple cookies for one site like this `cookie1=value1; [Domain=domain1; Path=path1;] cookie2=value2; [Domain=domain1; Path=path2;]`

You can also set headers with `RV_MEDIA_LIBRARY_UNSECURE_DEMO_HEADERS`. The format is the standard header format `key1=value1;key2=value2`

Finally, you can tell the demo plugin to download the file on disk and redirect the FileSourceIPNode to it with `RV_MEDIA_LIBRARY_UNSECURE_DEMO_FORCE_LOCAL_CACHE` set to 1. 
 
#### Debugging the plugins

You can debug what gets parsed and sent to ffmpeg with `RV_DEBUG_FFMPEG_COOKIES` and `RV_DEBUG_FFMPEG_HEADERS`.

#### EventMediaLibrary

The EventMediaLibrary got migrated to a plugin so we don't have two implementations to maintain. 

### Describe the reason for the change.

RVIO cannot use the EventMediaLibrary because RV Modes are not loaded. If your integration relies on being able to provide cookies or headers to RV/RVIO to play and export from media stored on the web, it only works in RV, and not in RVIO. 

The MediaLibrary plugins are always loaded, unless you set `RV_USE_MEDIA_LIBRARY_PLUGINS` to 0. 

### Describe what you have tested and on which operating system.

I tested it on MacOS, but nothing is really platform-specific. 

### Add a list of changes, and note any needing special attention during the review.

See pull request comments. 

### If possible, provide screenshots.

N/A

### CODE REVIEW

File: src/lib/app/PyTwkApp/PyInterface.cpp [hunk 1]
```diff
@@ -266,7 +267,16 @@ namespace TwkApp
 
     PySys_SetArgvEx( argc, w_argv, 0 );
 
-    PyLockObject::initialize();
+    //
```
Comment (geffrak) [inline src/lib/app/PyTwkApp/PyInterface.cpp:270]: This wasn't PyLockObject-related...

File: src/lib/app/RvCommon/RvApplication.cpp [hunk 2]
```diff
@@ -244,7 +244,7 @@ static void setEnvVar(const string& var, const string& val)
 
 RvApplication::RvApplication(int argc, char** argv)
     : QObject(),
-      IPCore::Application(),
+      RvConsoleApplication(),
```
Comment (geffrak) [inline src/lib/app/RvCommon/RvApplication.cpp:247]: An RvApplication is a RvConsoleApplication that is an IPCore::Application. The RVConsoleApplication is just for the PyMediaLibrary, and it's not a QObject

File: src/lib/ip/IPBaseNodes/FileSourceIPNode.cpp [hunk 3]
```diff
@@ -2680,9 +2680,10 @@ namespace IPCore
             for( int c = 0; c < cookies.size(); c++ )
             {
               if( c > 0 ) cookieStm << "\n";
-              cookieStm << cookies[c].name << "=" << cookies[c].value
-                        << "; path=" << cookies[c].path
-                        << "; domain=" << cookies[c].domain;
+              cookieStm << cookies[c].name << "=" << cookies[c].value;
+
+              if (!cookies[c].path.empty()) cookieStm << "; path=" << cookies[c].path;
```
Comment (geffrak) [inline src/lib/ip/IPBaseNodes/FileSourceIPNode.cpp:2685]: A cookie without a path and a domain is a valid cookie.

File: src/lib/mu/MuPy/MuPy/PyModule.h [hunk 4]
```diff
@@ -11,34 +11,6 @@
 #include <Mu/Value.h>
 
 #include <Python.h>
-class PyLockObject
```
Comment (geffrak) [inline src/lib/mu/MuPy/MuPy/PyModule.h:14]: Moved outside of this Mu library since it's required elsewere.

**Review by bernie-laberge — APPROVED** [review]
VERY Impressive work Kerby ! Superb !

**Comment by geffrak** [comment]
@mattdaw, at some point, you requested to be able to create MediaLibraries from Python Plugins. Now you can.

**Comment by mattdaw** [comment]
@geffrak Nice! Thanks for letting me know

**Review by cedrik-fuoco-adsk — APPROVED** [review]

**Pull request merged**